### PR TITLE
[WIP] Refactors the Holodeck to use Map Templates

### LINKED ISSUE #41485: Holodeck matchboxes drop matches on deletion
`/obj/item/storage/box/matches`

When holodeck matchboxes get deleted, they will drop their contents as regular items that are missing the `HOLOGRAM_1` flag.

### LINKED ISSUE #49318: Holodeck is a Nar'Sie summoning site but the veil is not weak enough to draw the rune
Issue reported from  Round ID: 130099 (/tg/Station Event Hall: Sseth overflow event. [ENGLISH] [EU] [100% FREE LAG])

Reporting client version: 513.1510

Read title.

### LINKED ISSUE #49673: Holodeck blood unwashable
Issue reported from  Round ID: 131367 (/tg/Station Manuel [ENGLISH] [US-WEST] [ROLEPLAY] [99% LAG FREE])

Reporting client version: 513.1507

-activate winter wonderland module
-kill penguin for BLOOD
-become bloody
-shower and stay messy(soap works fine apparently)
-cry when your colleagues make fun of you

### PR DESCRIPTION
**Pull request by Kylerace**
Screenshot_158

## About The Pull Request
Refactors the holodeck to use map templates instead of copy_contents_to, which every maintainer seems to have complaints about. 
Fixes #41485 due to an added GetAllContents() proc for all containers in the spawned list
Fixes #49318 because the holodeck doesnt actually change areas anymore, and holodeck/rec_center has the VALID_TERRITORY flag
Fixes #49673, i dont know if this was fixed in some earlier pr but the behavior it talks about doesnt happen anymore so im counting it

## Why It's Good For The Game
The holodeck sims wont take up space in the centcom z level any more (which allows for more possible programs in the future), and map templates are more heavily tested. This is also a chance to future proof the holodeck against bugs if possible. Holodeck also seems more responsive. This should allow for a second custom holodeck in some future ruin as well, e.g. var edited for the offline program to look different and/or a completely different list of programs

## Changelog
:cl:

refactor: Nanotrasen overhauled much of the holodecks programming so that it didnt have to keep physical copies of the simulations anymore.
code: holodeck uses map templates instead of area_copy
/:cl:
EDIT: a lot of this following paragraph is outdated, the list of atoms created by the holodeck is now returned by InitializeAtoms inside of SSatoms, parsed_map/initTemplateBounds passes an argument to InitializeAtoms that makes it return a list only if the map template itself is set to return a list. initAtom() adds itself to the return list in SSatoms if the subsystem is set to return one (it unsets itself after the template is done, and i wasnt able to find a way to trick SSatoms into adding a non template atom to the list when its not supposed to, but it might be possible).
Right now im trying to cut down the number of for loops that iterate through every turf in the holodeck and/or its contents, as well as making sure that my changes to the vanilla map template and parsed_map code doesnt fuck up something other than holodeck.
 i'll write this more fully tomorrow but basically datum/map_template and datum/parsed map both have an added list 
called created_atoms and a boolean returns_list that defaults to false. when template.load() is called it sets its parsed_map instance to share the same boolean and list, then inside parsed_map/create_atom() it adds the created atom to the list if returns_list is true. the flaws ive found with this approach so far is 
1. this by itself doesnt work for objects that create other objects on initialization (like courtroom lockers), requiring all items in spawned to be checked for any contents. 
2.  is a bit ugly since it requires the new values to be passed down and then up the daisy chain of holodeck -> map_template instance -> parsed_map instance (returns_list is inherited from map_template, list is created here) -> same map_template instance (now has populated list) -> holodeck. However map templates already have a similar daisy chain except in the form of proc arguments instead of variables assigned at creation.

### CODE REVIEW

**Comment by optimumtact** [comment]
remove the change log changes you have made and disable changelog CI on your fork of tgstation.

**Comment by LemonInTheDark** [comment]
You've got some file paths that have been updated, and it looks like you need to rebuild tgui, but this is promising, it's been a long time coming too, godspeed :)

**Comment by LemonInTheDark** [comment]
Oh also add more detail to what was changed in the changelog

**Comment by Fikou** [comment]
you forgot to delete the holodecks from centcom

File: code/modules/holodeck/computer.dm
Comment (Dennok) [inline code/modules/holodeck/computer.dm:72]: I think that `!offline_program` must be in separated check and log something like `Wrong offline_program variable`

File: code/modules/holodeck/computer.dm
Comment (Dennok) [inline code/modules/holodeck/computer.dm:59]: z = 2 ?
It maybe place here cumputer z?

File: code/modules/holodeck/computer.dm
Comment (Dennok) [inline code/modules/holodeck/computer.dm:82]: until someone want custom offlike holodeck, like space, or plasma burn range or etc.

File: code/modules/holodeck/computer.dm
Comment (Dennok) [inline code/modules/holodeck/computer.dm:399]: What is the reason for this?

File: code/modules/holodeck/turfs.dm
Comment (Dennok) [inline code/modules/holodeck/turfs.dm:5]: Why not just use dir?

File: code/modules/mapping/map_template.dm
Comment (Dennok) [inline code/modules/mapping/map_template.dm:111]: This one make block of turfs between lower left-1 and upper right+1. outer border

File: tgui/packages/tgui/interfaces/Holodeck.js
Comment (Dennok) [inline tgui/packages/tgui/interfaces/Holodeck.js:34]: This change really matters?

File: code/modules/mapping/map_template.dm
Comment (Kylerace) [reply code/modules/mapping/map_template.dm:111]: I wasn't really sure what key did but I don't pass program.type to tgui anymore so I changed it to id

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: fixed, now uses the offline_program var again in case someone wants to var edit a custom holodeck

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: fixed, now its bottom_left = locate(linked.x, linked.y, src.z)

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: fixed

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: I was probably playing it safe, fixed

**Comment by Kylerace** [comment]
Fuck the changelog, all my homies hate the changelog

**Comment by Dennok** [comment]
You can place area noop  to templates. To make it not change area.

**Comment by Kylerace** [comment]
> You can place area noop to templates. To make it not change area.
So the holodeck should have one area no matter what program is loaded? @Dennok

**Comment by Dennok** [comment]
> > You can place area noop to templates. To make it not change area.
> 
> So the holodeck should have one area no matter what program is loaded? @Dennok

Yes

File: code/modules/holodeck/holodeck.dm
Comment (Cyberboss) [inline code/modules/holodeck/holodeck.dm:1]: Name this `map_templates.dm` or something else more descriptive

File: code/modules/holodeck/turfs.dm
Comment (Cyberboss) [inline code/modules/holodeck/turfs.dm:58]: Delete this proc

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/holodeck/holodeck.dm
Comment (Kylerace) [reply code/modules/holodeck/holodeck.dm]: now called holodeck_map_templates and updated comment at the top of computer.dm

File: code/modules/holodeck/turfs.dm
Comment (Kylerace) [reply code/modules/holodeck/turfs.dm]: done

**Comment by Kylerace** [comment]
need to fix explosion bug, after that its good to go

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:160]: If this is the only change it would be better to make this variable on map_template type.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:185]: Same thing here. Duplication here is bad.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:173]: Is this necessary if you already iterate them one by one during template initialization.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm:173]: this is for lockers that create other items on initialization, thus the insides arent actually spawned from the template which makes them untracked otherwise

Comment (Kylerace) [reply code/modules/holodeck/computer.dm:173]: ill probably end up putting this part into load_program

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm:162]: removed snowflake holodeck map loader procs, now all map_templates (and parsed_map) instances have a list and boolean var that defaults to false, if its set to true then parsed_map/proc/create_atom() adds the created element to the list.

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm:187]: no more duplication

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm:175]: leaving this open because its possible theres a better way to accomplish this without manually adding all contents of every spawned container just in case they create items on Initialize()

**Comment by LemonInTheDark** [comment]
need more review?

**Comment by Kylerace** [comment]
> need more review?

@LemonInTheDark yes reviews would help a lot. specifically I need to know if what im doing in map_template.dm and reader.dm (plus the map_template/holodeck/update_blacklist() proc in line 135 of computer.dm) is ok. I tested it and on average my branch and master are done with creating the station + lavaland at nearly the same time. if im right non holodeck map templates shouldnt be doing any more work than evaluating some if statements to be false and calling a proc that just returns

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: as of right now this is fixed unless i get smacked for altering ssatoms to return a list of atoms created from a template (if the template is set to return a list, which only holodeck templates are currently)

**Comment by AnturK** [comment]
How's this going.

**Comment by Kylerace** [comment]
@AnturK ive decided to shelve the holo element idea since it doesnt help enough to be worth it. Might return to it in a future pr though. Currently I don't see anything else to do with it unless a maintainer wants me to change something.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:126]: You seem to have dropped the check for if the template actually exists and if emagged template is loaded into actually emagged holodeck.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:182]: Description in comments what's happening here please.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:178]: If you're doing istype here you can just do typed loop, it's more readable.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:201]: Would be nice to wrap these holoeffects into some typed proc while you're doing this.

File: code/modules/holodeck/computer.dm
Comment (AnturK) [inline code/modules/holodeck/computer.dm:233]: Why is this move necessary here ?

File: code/modules/mapping/reader.dm
Comment (AnturK) [inline code/modules/mapping/reader.dm:394]: Remains here and in vars

**Review by AnturK — CHANGES_REQUESTED** [review]
Looks mostly okay, a few minor issues. Also needs docs.

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm:233]: leftover from original holodeck code that i didnt get rid of, ill take it out in the next update (assuming it doesnt ruin anything but i dont see why it would)

**Comment by Kylerace** [comment]
definitely dont merge yet, have consistent oom crashing issues since one or two commits ago

File: code/modules/holodeck/computer.dm
Comment (Kylerace) [reply code/modules/holodeck/computer.dm]: found out why its necessary, if a non holodeck item is inside of a holodeck container (like the morgue trays), they can get qdel'd with the container without this proc. so far it only seems to work like that with the morgue tray, I assume that closets do it themselves on deletion.

**Comment by Kylerace** [comment]
havent been able to make real progress on this in days, verifying that a given version doesnt cause occasional crashes takes forever. I have to shelve this until next quarter in two weeks or so, i have too many things going on right now unfortunately.

**Pull request closed without merging**
